tensordict/tensorclass.py

```diff
@@ -856,20 +910,34 @@ def __torch_function__(
     cls.fields = classmethod(dataclasses.fields)
     for field in cls.fields():
         if hasattr(cls, field.name):
-            delattr(cls, field.name)
-
-    _get_type_hints(cls)
-    cls.__init__ = _init_wrapper(cls.__init__, frozen, shadow)
+            # if we have used Cls(TensorClass["shadow"]), we have a subclass of Cls(TensorClass)
+            #  so we cannot directly delete the attribute
+            try:
+                delattr(cls, field.name)
+            except AttributeError:
+                pass
+    _get_type_hints(cls, tensor_only=tensor_only)
+    if tensor_only and (cls._autocast or cls._nocast):
+        raise TypeError("tensor_only and _autocast or _nocast are exclusive features.")
+    cls.__init__ = _init_wrapper(cls.__init__, cls, frozen, shadow, tensor_only)
     cls._from_tensordict = classmethod(_from_tensordict)
     cls.from_tensordict = cls._from_tensordict
     if not hasattr(cls, "__torch_function__"):
         cls.__torch_function__ = classmethod(__torch_function__)
     cls.__getstate__ = _getstate
     cls.__setstate__ = _setstate
 
-    cls.__getattr__ = _getattr
+    if tensor_only:
+        cls.__getattr__ = _getattr_tensor_only
+    else:
+        cls.__getattr__ = _getattr
```

Comment:
very minor nit: as this is short the condition can be only on the value and not the assignment. Feel free to ignore
```suggestion
    cls.__getattr__ = _getattr_tensor_only if tensor_only else _getattr
```